internal/app/connectconformance/server_runner.go

```diff
@@ -26,8 +26,145 @@ import (
+		conformancev1alpha1.StreamType_STREAM_TYPE_SERVER_STREAM:
+
+		// First, find the response definition that the client instructed the
+		// server to return
+		var def *conformancev1alpha1.StreamResponseDefinition
+		if testCase.Request.StreamType == conformancev1alpha1.StreamType_STREAM_TYPE_SERVER_STREAM {
+			ur := &conformancev1alpha1.ServerStreamRequest{}
+			err := req.UnmarshalTo(ur)
+			if err != nil {
+				return err
+			}
+			def = ur.ResponseDefinition
+		} else {
+			ur := &conformancev1alpha1.BidiStreamRequest{}
+			err := req.UnmarshalTo(ur)
+			if err != nil {
+				return err
+			}
+			def = ur.ResponseDefinition
+		}
+
+		// Server should have echoed back all specified headers, trailers, and errors
+		expected := &conformancev1alpha1.ClientResponseResult{
+			ResponseHeaders:  def.ResponseHeaders,
+			ResponseTrailers: def.ResponseTrailers,
+			Error:            def.Error,
+		}
+
+		// There should be one payload for every ResponseData the client specified
+		expected.Payloads = make([]*conformancev1alpha1.ConformancePayload, len(def.ResponseData))
+
+		for idx, data := range def.ResponseData {
+			expected.Payloads[idx] = &conformancev1alpha1.ConformancePayload{
+				Data: data,
+			}
+			switch testCase.Request.StreamType { //nolint:exhaustive
+			case conformancev1alpha1.StreamType_STREAM_TYPE_SERVER_STREAM:
+				// For server streams, only one request was sent. All request information specified
+				// should only be echoed back in the first response
+				if idx == 0 {
+					expected.Payloads[idx].RequestInfo = &conformancev1alpha1.ConformancePayload_RequestInfo{
+						RequestHeaders: testCase.Request.RequestHeaders,
+						Requests:       testCase.Request.RequestMessages,
+					}
+				}
+			case conformancev1alpha1.StreamType_STREAM_TYPE_HALF_DUPLEX_BIDI_STREAM:
+				// For half-duplex streams, all request information should be echoed back in all
+				// responses. For example, if 3 requests were sent and we received 4 responses, the information
+				// for all 3 requests should be present in all 4 payload responses.
+				expected.Payloads[idx].RequestInfo = &conformancev1alpha1.ConformancePayload_RequestInfo{
+					RequestHeaders: testCase.Request.RequestHeaders,
+					Requests:       testCase.Request.RequestMessages,
+				}
+			case conformancev1alpha1.StreamType_STREAM_TYPE_FULL_DUPLEX_BIDI_STREAM:
+				// For a full duplex stream, the first request information should be echoed back in the first
+				// payload. The second should be echoed back in the second payload, etc. (i.e. a ping pong interaction)
+				expected.Payloads[idx].RequestInfo = &conformancev1alpha1.ConformancePayload_RequestInfo{
+					RequestHeaders: testCase.Request.RequestHeaders,
+					Requests:       []*anypb.Any{testCase.Request.RequestMessages[idx]},
+				}
+			}
+		}
+		testCase.ExpectedResponse = expected
+	case conformancev1alpha1.StreamType_STREAM_TYPE_UNARY, conformancev1alpha1.StreamType_STREAM_TYPE_CLIENT_STREAM:
```

Comment:
These cases are big enough and reasonably self-contained, so I think it would be better to factor them out into two methods -- like a `populateExpectedStreamResponse` and `populateExpectedUnaryResponse`.

Reply:
Moved out into their own functions, but I kept the entry point for `populateExpectedResponse` so it didn't clutter up the `runTestCasesForServer` function. I can remove that also too, so just lmk.